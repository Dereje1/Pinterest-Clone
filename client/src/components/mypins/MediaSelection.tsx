import React from 'react';
import DriveFolderUploadIcon from '@mui/icons-material/DriveFolderUpload';
import ToggleButton from '@mui/material/ToggleButton';
import ToggleButtonGroup from '@mui/material/ToggleButtonGroup';
import LinkIcon from '@mui/icons-material/Link';
import PsychologyIcon from '@mui/icons-material/Psychology';
import RestartAltIcon from '@mui/icons-material/RestartAlt';
import Typography from '@mui/material/Typography';
import UploadFileIcon from '@mui/icons-material/UploadFile';
import IconButton from '@mui/material/IconButton';
import Button from '@mui/material/Button';
import TextField from '@mui/material/TextField';

interface MediaSelectProps {
    mediaType: 'link' | 'upload' | 'AI'
    totalAiGenratedImages: number
    handleMediaChange: (e: React.SyntheticEvent, imgType: 'link' | 'upload' | 'AI') => void
  }

export function MediaSelect({
  mediaType,
  totalAiGenratedImages,
  handleMediaChange,
}: MediaSelectProps) {
  return (
    <ToggleButtonGroup
      value={mediaType}
      exclusive
      onChange={handleMediaChange}
      aria-label="text alignment"
      sx={{ marginRight: 80, color: '900' }}
    >
      <ToggleButton value="link" aria-label="link type">
        <LinkIcon />
      </ToggleButton>
      <ToggleButton value="upload" aria-label="upload type">
        <DriveFolderUploadIcon />
      </ToggleButton>
      <ToggleButton value="AI" aria-label="AI type" disabled={totalAiGenratedImages >= 5}>
        <PsychologyIcon />
      </ToggleButton>
    </ToggleButtonGroup>
  );
}

interface MediaButtonHandlerProps {
    mediaType: 'link' | 'upload' | 'AI'
    totalAiGenratedImages: number
    isError: boolean
    picPreview: string | ArrayBuffer
    description: string
    AIimageStatus: {
        generatedImage: boolean,
        generatingImage: boolean,
        _id: string | null
    }
    handleLinkImage: (e: React.SyntheticEvent) => void
    handleUploadedImage: (e: React.SyntheticEvent<HTMLDivElement>) => void
    handleAIimage: () => void
    showAIResetDialog: () => void
  }

export function MediaButtonHandler({
  mediaType,
  isError,
  picPreview,
  description,
  totalAiGenratedImages,
  AIimageStatus,
  handleLinkImage,
  handleUploadedImage,
  handleAIimage,
  showAIResetDialog,
}: MediaButtonHandlerProps) {
  const handler = {
    upload: (
      <Button
        variant="contained"
        startIcon={<UploadFileIcon />}
        sx={{ margin: '2.3vh' }}
        component="label"
        color={isError ? 'error' : 'secondary'}
      >
        {isError ? 'choose image' : 'replace image'}
        <input hidden accept="image/*" type="file" onChange={handleUploadedImage} />
      </Button>
    ),
    link: (
      <TextField
        id="pin-img-link"
        label="Paste image address here http://..."
        variant="standard"
        onChange={handleLinkImage}
        value={picPreview}
        error={isError}
        color="success"
        style={{ margin: '1.5vh', width: '100%' }}
      />
    ),
    AI: (
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <Button
          variant="contained"
          startIcon={<PsychologyIcon />}
          sx={{ margin: '2.3vh' }}
          component="label"
          color="info"
          onClick={handleAIimage}
          disabled={
            AIimageStatus.generatedImage
                || AIimageStatus.generatingImage
                || !description.trim().length
                || totalAiGenratedImages >= 5
          }
        >
          {AIimageStatus.generatedImage ? 'Generated' : 'Generate Image'}
        </Button>
        <IconButton
          onClick={showAIResetDialog}
          disabled={
            !AIimageStatus.generatedImage
                || AIimageStatus.generatingImage
                || totalAiGenratedImages >= 5
          }
          disableRipple
        >
          <RestartAltIcon />
        </IconButton>
        <Typography>
          {totalAiGenratedImages}
          /5
        </Typography>
      </div>
    ),
  };

  return handler[mediaType];
}
