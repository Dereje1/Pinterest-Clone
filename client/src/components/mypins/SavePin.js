import React from 'react';
import PropTypes from 'prop-types';
import CardActions from '@mui/material/CardActions';
import SaveAltIcon from '@mui/icons-material/SaveAlt';
import IconButton from '@mui/material/IconButton';
import TextField from '@mui/material/TextField';

const SavePin = ({ isImageError, isDescriptionError, savePic }) => {
  let validation;
  if (isImageError && isDescriptionError) {
    validation = {
      text: 'Invalid image and description',
      color: '#f79f9fd9',
    };
  } else if (isImageError) {
    validation = {
      text: 'Invalid image',
      color: '#f79f9fd9',
    };
  } else if (isDescriptionError) {
    validation = {
      text: 'Invalid description',
      color: '#f79f9fd9',
    };
  } else {
    validation = {
      text: 'Save pin',
      color: '#028b02',
    };
  }
  return (
    <CardActions
      disableSpacing
    >
      <IconButton
        aria-label="Pin Image"
        onClick={savePic}
        disabled={isImageError || isDescriptionError}
      >
        <SaveAltIcon style={{ fontSize: '1.5em', color: validation.color }} />
      </IconButton>
      {
        (isImageError || isDescriptionError) && (
          <TextField
            id="pin-create-validation"
            value={validation.text}
            variant="standard"
            sx={{
              marginTop: 2,
              width: '100%',
              input: {
                fontSize: '0.75em',
                color: validation.color,
                fontWeight: 'bold',
                caretColor: 'white',
              },
            }}
            InputProps={{
              disableUnderline: true,
            }}
          />
        )
      }

    </CardActions>
  );
};

export default SavePin;

SavePin.propTypes = {
  isImageError: PropTypes.bool.isRequired,
  isDescriptionError: PropTypes.bool.isRequired,
  savePic: PropTypes.func.isRequired,
};
