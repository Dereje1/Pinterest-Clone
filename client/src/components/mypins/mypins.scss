
#mypinframe{
    margin-top: 5px;
    margin-left: 20px;
    margin-right: 20px;
    padding: 0px;
    display: flex;
    flex-direction: column;
    align-items: center;
    #username{
      font-weight: 700;
      font-family: "Helvetica";
      font-size: 36px;
      letter-spacing: -1.25px;
      line-height: 0px;
      word-spacing: -2.25px;
      color: #555;
    }
    #creatpinwrapper{
      margin: 0 auto;
      display: flex;
      flex-direction: column;
      background: rgba(0, 0, 0, 0.05);
      border-radius: 8px;
      padding: 10px;
      visibility: hidden;
      cursor: pointer;
      @media (hover: hover) {
        &:hover{
          visibility: visible;
        }
      }
      #createpin{
        height:200px;
        width:301px;
        display: flex;
        justify-content: center;
        align-items: center;
        background: rgba(0, 0, 0, 0.05);
        border-radius: 8px;
        visibility: visible;
        .fa-plus-circle{
          color:red;
          font-size:50px;
        }
      }
      #createpintext{
        color:#b5b5b5;
        font-size: 18px;
        visibility: visible;
        font-weight: 700;
        font-family: Helvetica;
      }
    }
  }