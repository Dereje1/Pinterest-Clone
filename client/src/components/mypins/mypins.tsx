// user page only for authenticated users
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Avatar from '@mui/material/Avatar';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogTitle from '@mui/material/DialogTitle';
import Box from '@mui/material/Box';
import Fab from '@mui/material/Fab';
import AddIcon from '@mui/icons-material/Add';
import PinCreate from './pincreatemodal';
import ImageBuild from '../imagebuild/Imagebuild';
import RESTcall from '../../crud'; // pin CRUD
import { getProviderIcons, Loading, UserPinsSelector } from '../common/common';
import {
  providerIconsType, userType, PinType,
} from '../../interfaces';

const providerIcons = getProviderIcons({ fontSize: 45 });

interface getUserInfoProps {
  service: string
  displayName: string | null
  username: string | null
}
const getUserInfo = ({ service, displayName, username }: getUserInfoProps) => (
  <div style={{
    display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'center',
  }}
  >
    <Avatar sx={{
      mb: 2,
      width: 52,
      height: 52,
      bgcolor: providerIcons[service as keyof providerIconsType].color,
    }}
    >
      {providerIcons[service as keyof providerIconsType].icon}
    </Avatar>
    <Typography variant="h4">
      {displayName}
    </Typography>
    <Typography style={{ marginLeft: 15 }}>
      {service === 'google' ? username : `@${username}`}
    </Typography>
  </div>
);

interface MypinsProps {
  user: userType
}

interface MypinsState {
  displayPinCreate: boolean
  pinList: PinType[]
  showDeleteImageModal: boolean
  deletableImgInfo: PinType | null,
  ready: boolean,
  displaySetting: string,
}

export class Mypins extends Component<MypinsProps, MypinsState> {

  constructor(props: MypinsProps) {
    super(props);
    this.state = {
      displayPinCreate: false, // controls pin creation modal
      pinList: [], // collects the pins user owns and saved
      showDeleteImageModal: false,
      deletableImgInfo: null,
      ready: false,
      displaySetting: 'created',
    };
  }

  async componentDidMount() {
    // get all pins and filter by owned and saved and then concatenate and set state
    const { profilePins } = await RESTcall({
      address: '/api/mypins',
      method: 'get',
      payload: undefined,
    });
    this.setState({
      pinList: profilePins,
      ready: true,
    });
  }

  // adds a new pin to the db
  async addPic(pinJSON: {imgDescription: string, imgLink: string | ArrayBuffer}) {
    // copy then add pin to db and then update client state (in that order)
    const { pinList } = this.state;
    this.setState({ ready: false });
    const newPin = await RESTcall({
      address: '/api/newpin',
      method: 'post',
      payload: pinJSON,
    });
    const addedPin = {
      ...newPin,
      owns: true,
      hasSaved: false,
    };
    this.setState({
      pinList: [...pinList, addedPin],
      ready: true,
      displayPinCreate: false,
      displaySetting: 'created',
    });
  }

  pinForm() { // display pin creation modal
    this.setState({
      displayPinCreate: true,
    });
  }

  deletePic({ _id, owns }: {_id: string, owns: boolean}) {
    const { pinList, displayPinCreate } = this.state;
    if (displayPinCreate) return;
    this.setState({
      pinList: pinList.filter((p) => p._id !== _id),
      showDeleteImageModal: false,
      deletableImgInfo: null,
    }, async () => {
      await RESTcall({
        address: owns ? `/api/${_id}` : `/api/unpin/${_id}`,
        method: owns ? 'delete' : 'put',
        payload: undefined,
      });
    });
  }

  render() {
    const { user, user: { authenticated } } = this.props;
    const {
      displayPinCreate, showDeleteImageModal,
      deletableImgInfo, pinList,
      ready, displaySetting,
    } = this.state;
    if (!authenticated) return null;
    if (!ready) return <Loading marginTop={100} />;

    const pins = pinList.filter((pin) => (displaySetting === 'created' ? Boolean(pin.owns) : Boolean(pin.hasSaved)));
    return (
      <>
        <div style={{
          marginTop: 80,
          marginLeft: 'auto',
          marginRight: 'auto',
          width: 300,
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'space-between',
        }}
        >
          {getUserInfo(user)}
          <div style={{ marginRight: 18 }}>
            <UserPinsSelector
              displaySetting={displaySetting}
              setDisplaySetting={(val) => this.setState({ displaySetting: val })}
            />
          </div>

          <Fab
            id="createpin"
            color="default"
            aria-label="add"
            onClick={() => this.pinForm()}
            sx={{ marginTop: 1, zIndex: 1 }}
          >
            <AddIcon fontSize="large" />
          </Fab>
        </div>
        {
          pins.length
            ? (
              <ImageBuild
                pinImage={false}
                deletePin={(pin) => {
                  if (pin.owns) {
                    this.setState({
                      showDeleteImageModal: true,
                      deletableImgInfo: pin,
                    });
                  } else {
                    this.deletePic(pin);
                  }
                }}
                pinList={pins}
                ready={ready}
                displayBrokenImage
                user={user}
              />

            )
            : (
              <Box
                sx={{
                  marginLeft: 'auto',
                  marginRight: 'auto',
                  marginTop: 10,
                  padding: 2,
                  backgroundColor: '#eeeeee',
                }}
              >
                { displaySetting === 'saved' ? (
                  <Typography variant="h6" textAlign="center">
                    You have no pins saved, you can save some from the
                    {' '}
                    <Link to="/">home</Link>
                    {' '}
                    page
                  </Typography>
                )
                  : (
                    <>
                      <Typography variant="h4" style={{ textAlign: 'center' }}>
                        Welcome!
                      </Typography>
                      <Typography variant="h6" textAlign="center">
                        You have not created any pins, you can create some
                        by clicking on the button above
                      </Typography>
                    </>
                  )}
              </Box>
            )
        }
        {displayPinCreate && (
          <PinCreate
            reset={() => this.setState({ displayPinCreate: false })}
            savePin={(pinJSON) => this.addPic(pinJSON)}
          />
        )}
        { showDeleteImageModal && (
          <Dialog
            open={showDeleteImageModal}
            aria-labelledby="alert-dialog-title"
            aria-describedby="alert-dialog-description"
          >
            <DialogTitle id="alert-dialog-title">
              {`Permanently delete ${deletableImgInfo && deletableImgInfo.imgDescription}?`}
            </DialogTitle>
            <DialogActions>
              <Button
                id="cancel-delete-alert"
                onClick={() => this.setState({
                  showDeleteImageModal: false,
                  deletableImgInfo: null,
                })}
              >
                Cancel
              </Button>
              <Button
                id="resume-delete-alert"
                onClick={() => {
                  if (deletableImgInfo !== null) {
                    this.deletePic(deletableImgInfo);
                  }
                }}
                autoFocus
              >
                Delete
              </Button>
            </DialogActions>
          </Dialog>
        )}
      </>
    );
  }

}

export const mapStateToProps = ({ user }:{user: userType}) => ({ user });

export default connect(mapStateToProps)(Mypins);
