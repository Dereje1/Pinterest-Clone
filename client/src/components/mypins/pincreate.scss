.pincreate{
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    margin-left: auto;
    margin-right: auto;
    background: white;
    box-shadow: 0 5px 15px rgba(0,0,0,.5);
    border: 1px solid rgba(0,0,0,.2);
    border-radius: 6px;
    z-index: 1010;
    &.cshow{
      display: block;
      animation-name: modalshow;
      animation-duration: 1s;
      animation-timing-function: cubic-bezier(0.075, 0.82, 0.165, 1);
      animation-iteration-count: 1;
    }
    &.chide{
      display: block;
      visibility: hidden;
      animation-name: modalhide;
      animation-duration: 1s;
      animation-timing-function: cubic-bezier(0.075, 0.82, 0.165, 1);
      animation-iteration-count: .5;
    }
  }
  
  @keyframes modalshow {
    0%{
      opacity: 0;
      transform: scale(0)
    }
    100%{
      opacity: 1;
      transform: scale(1)
    }
  }
  
  @keyframes modalhide {
    0%{
      transform: scale(1);
      visibility: visible;
    }
    100%{
      transform: scale(0);
    }
  }