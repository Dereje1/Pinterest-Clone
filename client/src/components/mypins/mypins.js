// user page only for authenticated users
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import TwitterIcon from '@mui/icons-material/Twitter';
import GoogleIcon from '@mui/icons-material/Google';
import GitHubIcon from '@mui/icons-material/GitHub';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogTitle from '@mui/material/DialogTitle';
import Box from '@mui/material/Box';
import PinCreate from './pincreatemodal';
import ImageBuild from '../imagebuild/Imagebuild';
import Loading from '../imagebuild/Loading';
import RESTcall from '../../crud'; // pin CRUD
import './mypins.scss';

const getUserName = ({ service, displayName, username }) => {
  const serviceObj = {
    twitter:
  <>
    <TwitterIcon style={{ fontSize: 30, color: '#1DA1F2' }} />
    <Typography style={{ marginLeft: 15 }}>
      {`@${username}`}
    </Typography>
  </>,
    google:
  <>
    <GoogleIcon style={{ fontSize: 30, color: '#4285F4' }} />
    <Typography style={{ marginLeft: 15 }}>
      {username}
    </Typography>
  </>,
    github:
  <>
    <GitHubIcon style={{ fontSize: 30, color: '#1d7b20' }} />
    <Typography style={{ marginLeft: 15 }}>
      {`@${username}`}
    </Typography>
  </>,
  };
  return (
    <div style={{
      marginTop: 80,
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    }}
    >
      <Typography variant="h4">
        {displayName}
      </Typography>
      <div style={{ display: 'flex', alignItems: 'center' }}>
        {serviceObj[service]}
      </div>
    </div>
  );
};

export class Mypins extends Component {

  constructor(props) {
    super(props);
    this.state = {
      displayPinCreate: false, // controls pin creation modal
      pinList: [], // collects the pins user owns and saved
      allPinLinks: [], // URL of all pins in DB
      showDeleteImageModal: false,
      deletableImgInfo: null,
      ready: false,
    };
  }

  async componentDidMount() {
    // get all pins and filter by owned and saved and then concatenate and set state
    const { profilePins, allPinLinks } = await RESTcall({
      address: '/api/?type=profile',
      method: 'get',
    });
    this.setState({
      pinList: profilePins,
      allPinLinks,
      ready: true,
    });
  }

  async addPic(pinJSON) { // adds a pin to the db
    // copy then add pin to db and then update client state (in that order)
    const { pinList, allPinLinks } = this.state;
    this.setState({ ready: false });
    const newPin = await RESTcall({
      address: '/api/newpin',
      method: 'post',
      payload: pinJSON,
    });
    const { owner, imgLink, originalImgLink } = newPin;
    const addedPin = {
      ...newPin,
      owner: owner.name,
      owns: true,
      hasSaved: false,
    };
    this.setState({
      pinList: [...pinList, addedPin],
      allPinLinks: [...allPinLinks, { imgLink, originalImgLink }],
      ready: true,
      displayPinCreate: false,
    });
  }

  pinForm() { // display pin creation modal
    this.setState({
      displayPinCreate: true,
    });
  }

  deletePic({ _id, imgLink, owns }) {
    const { pinList, displayPinCreate, allPinLinks } = this.state;
    if (displayPinCreate) return;
    this.setState({
      pinList: pinList.filter(p => p._id !== _id),
      showDeleteImageModal: false,
      deletableImgInfo: null,
      allPinLinks: owns ? allPinLinks.filter(links => links.imgLink !== imgLink) : allPinLinks,
    }, async () => {
      await RESTcall({
        address: owns ? `/api/${_id}` : `/api/unpin/${_id}`,
        method: owns ? 'delete' : 'put',
      });
    });
  }

  render() {
    const { user, user: { authenticated, displayName } } = this.props;
    const {
      displayPinCreate, showDeleteImageModal,
      deletableImgInfo, allPinLinks, pinList,
      ready,
    } = this.state;
    if (!authenticated) return null;
    if (!ready) return <div style={{ marginTop: 100 }}><Loading /></div>;

    return (
      <React.Fragment>
        <div>
          <div id="mypinframe">
            {getUserName(user)}
            <div
              id="creatpinwrapper"
              onClick={() => this.pinForm()}
              role="button"
              onKeyDown={() => { }}
              tabIndex={0}
            >
              <div id="createpin">
                <i className="fa fa-plus-circle" aria-hidden="true" />
              </div>
              <h3 id="createpintext">Create Pin</h3>
            </div>
          </div>
          {
            pinList.length
              ? (
                <ImageBuild
                  pinImage={null}
                  deletePin={(e) => {
                    if (e.owns) {
                      this.setState({
                        showDeleteImageModal: true,
                        deletableImgInfo: e,
                      });
                    } else {
                      this.deletePic(e);
                    }
                  }}
                  pinList={pinList}
                  ready={ready}
                  displayBrokenImage
                  displayName={displayName}
                />
              )
              : (
                <Box
                  sx={{
                    marginLeft: 'auto',
                    marginRight: 'auto',
                    marginTop: 10,
                    padding: 2,
                    backgroundColor: '#eeeeee',
                  }}
                >
                  <Typography variant="h4" style={{ textAlign: 'center' }}>
                    Welcome!
                  </Typography>
                  <Typography variant="h6" textAlign="center">
                    To get started, you can create your own pin,
                    or save an existing pin from the
                    {' '}
                    <Link to="/">home</Link>
                    {' '}
                    page
                  </Typography>
                </Box>
              )
          }

          {displayPinCreate && (
            <PinCreate
              reset={() => this.setState({ displayPinCreate: false })}
              savePin={pinJSON => this.addPic(pinJSON)}
              allPinLinks={allPinLinks}
            />
          )}
          { showDeleteImageModal && (
            <Dialog
              open={showDeleteImageModal}
              aria-labelledby="alert-dialog-title"
              aria-describedby="alert-dialog-description"
            >
              <DialogTitle id="alert-dialog-title">
                {`Permanently delete ${showDeleteImageModal && deletableImgInfo.imgDescription}?`}
              </DialogTitle>
              <DialogActions>
                <Button
                  id="cancel-delete-alert"
                  onClick={() => this.setState({
                    showDeleteImageModal: false,
                    deletableImgInfo: null,
                  })}
                >
                  Cancel
                </Button>
                <Button id="resume-delete-alert" onClick={() => this.deletePic(deletableImgInfo)} autoFocus>
                  Delete
                </Button>
              </DialogActions>
            </Dialog>
          )}
        </div>
      </React.Fragment>
    );
  }

}

const mapStateToProps = state => state;

export default connect(mapStateToProps)(Mypins);

Mypins.defaultProps = {
  user: {},
};

Mypins.propTypes = {
  // authentication info from redux
  user: PropTypes.shape(PropTypes.shape),
};
