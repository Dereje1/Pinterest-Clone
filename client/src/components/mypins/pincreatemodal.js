// displays modal that creates pin
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CancelIcon from '@mui/icons-material/Cancel';
import TextField from '@mui/material/TextField';
import Card from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import CardActions from '@mui/material/CardActions';
import FormControlLabel from '@mui/material/FormControlLabel';
import UploadFileIcon from '@mui/icons-material/UploadFile';
import Switch from '@mui/material/Switch';
import IconButton from '@mui/material/IconButton';
import Button from '@mui/material/Button';
import DriveFolderUploadIcon from '@mui/icons-material/DriveFolderUpload';
import SavePin from './SavePin';
import error from './error.png';

import {
  delay,
  validateURL,
  getModalWidth,
  isDuplicateError,
  encodeImageFileAsURL,
} from '../../utils/utils';
import './pincreate.scss';


class PinCreate extends Component {

  constructor(props) {
    super(props);
    // initialize modal show state to false
    this.state = {
      picPreview: '', // on erroneous image links
      description: '',
      isError: true,
      isLoaded: false,
      upload: false,
    };
  }

  componentDidMount() {
    window.scrollTo(0, 0);
    this.setState({
      description: '',
      picPreview: '',
    }, async () => { await delay(1000); });
  }

  // disableScroll = () => window.scrollTo(0, 0);

  close = () => {
    const { reset } = this.props;
    this.setState({
      picPreview: '',
      description: '',
      isError: true,
    }, reset);
  };

  processImage = (e) => { // processes picture on change of text box
    const imgLink = validateURL(e.target.value);
    if (!imgLink) return;
    this.setState({
      picPreview: imgLink,
      isError: false,
      isLoaded: false,
    });
  };

  savePic = () => { // ready to save pin
    const { savePin } = this.props;
    const { picPreview, description } = this.state;
    // prepare JSON for POST api
    const pinJSON = {
      imgDescription: description,
      imgLink: picPreview,
    };
    // save into db and close modal
    savePin(pinJSON);
    this.close();
  };

  onError = () => {
    this.setState({
      picPreview: '',
      isLoaded: false,
      isError: true,
    });
  };

  onLoad = ({ target }) => {
    this.setState({
      isLoaded: true,
      isError: target.currentSrc === error,
    });
  };

  handleUploadedImage = async ({ target: { files } }) => {
    const [imgFile] = files;
    try {
      const ans = await encodeImageFileAsURL(imgFile);
      this.processImage({
        target: {
          value: ans,
        },
      });
    } catch (_) {
      this.onError();
    }
  };

  render() {
    const {
      description, isError, picPreview, isLoaded, upload,
    } = this.state;
    const { allPinLinks } = this.props;
    const modalWidth = getModalWidth();
    const isDescriptionError = description.trim().length < 5;
    const duplicateError = isDuplicateError(allPinLinks, picPreview);
    return (
      <>
        <div className="pin-create-modal-overlay" />
        <Card sx={{ width: modalWidth }} className="pincreate cshow">

          <CardHeader
            action={(
              <>
                <FormControlLabel
                  control={(
                    <Switch
                      checked={upload}
                      onChange={() => this.setState({ upload: !upload, picPreview: '' })}
                    />
                  )}
                  label={<DriveFolderUploadIcon />}
                  sx={{ marginRight: 8, color: '900' }}
                />
                <IconButton aria-label="settings" onClick={this.close}>
                  <CancelIcon style={{ fontSize: '1.5em', color: '#3a1c1cde' }} />
                </IconButton>
              </>
            )}
            title={`Create pin from ${upload ? 'file' : 'link'}`}
            titleTypographyProps={{ fontSize: '1.2em', fontWeight: 'bold' }}
          />
          <CardMedia
            component="img"
            image={picPreview === '' ? error : picPreview}
            onError={this.onError}
            onLoad={this.onLoad}
            sx={{ objectFit: 'contain', height: 250 }}
            id="new-pin-image"
          />
          <CardActions>
            <div style={{
              display: 'flex',
              flexDirection: 'column',
              justifyContent: 'space-evenly',
              alignItems: 'center',
              width: '100%',
            }}
            >
              <TextField
                id="pin-description"
                label="Description..."
                variant="standard"
                maxLength="28"
                color="success"
                onChange={({ target: { value } }) => value.trim().length <= 20
             && this.setState({ description: value })}
                value={description}
                error={!description || isDescriptionError}
                style={{ margin: '1.5vh', width: '100%' }}
              />
              {
                upload
                  ? (
                    <Button
                      variant="contained"
                      startIcon={<UploadFileIcon />}
                      sx={{ margin: '2.3vh' }}
                      component="label"
                      color={isError ? 'error' : 'secondary'}
                    >
                      {isError ? 'choose image' : 'replace image'}
                      <input hidden accept="image/*" type="file" onChange={this.handleUploadedImage} />
                    </Button>
                  )
                  : (
                    <TextField
                      id="pin-img-link"
                      label="Paste image address here http://..."
                      variant="standard"
                      onChange={e => this.processImage(e)}
                      value={picPreview}
                      error={isError}
                      color="success"
                      style={{ margin: '1.5vh', width: '100%' }}
                    />
                  )
              }
              <SavePin
                isImageError={isError || !isLoaded}
                isDescriptionError={isDescriptionError}
                isDuplicateError={duplicateError}
                savePic={this.savePic}
              />
            </div>
          </CardActions>

        </Card>
      </>
    );
  }

}

export default PinCreate;


PinCreate.propTypes = {
  // data used for pin creation
  allPinLinks: PropTypes.arrayOf(PropTypes.object).isRequired,
  // callback in mypins to turn modal off
  reset: PropTypes.func.isRequired,
  // POST request via axios
  savePin: PropTypes.func.isRequired,
};
