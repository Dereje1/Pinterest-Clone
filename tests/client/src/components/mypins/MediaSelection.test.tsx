import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MediaSelect, MediaButtonHandler } from '../../../../../client/src/components/mypins/MediaSelection';

describe('Selecting media', () => {
  test('will display the media selection toggle buttons', () => {
    const wrapper = shallow(<MediaSelect mediaType="link" handleMediaChange={jest.fn()} totalAiGenratedImages={2} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('Handling different media buttons -', () => {
  let props: React.ComponentProps<typeof MediaButtonHandler>;
  beforeEach(() => {
    props = {
      mediaType: 'link',
      isError: false,
      picPreview: '',
      description: 'stub description',
      totalAiGenratedImages: 2,
      AIimageStatus: {
        generatedImage: false,
        generatingImage: false,
        _id: null,
      },
      handleLinkImage: jest.fn(),
      handleUploadedImage: jest.fn(),
      handleAIimage: jest.fn(),
      showAIResetDialog: jest.fn(),
    };
  });
  test('links', () => {
    const wrapper = shallow(<MediaButtonHandler {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('uploads - with no error', () => {
    props = { ...props, mediaType: 'upload' };
    const wrapper = shallow(<MediaButtonHandler {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('uploads with error', () => {
    props = { ...props, mediaType: 'upload', isError: true };
    const wrapper = shallow(<MediaButtonHandler {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('AI - not generated yet', () => {
    props = { ...props, mediaType: 'AI' };
    const wrapper = shallow(<MediaButtonHandler {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('AI - generated ', () => {
    props = {
      ...props,
      mediaType: 'AI',
      AIimageStatus: {
        generatedImage: true,
        generatingImage: false,
        _id: null,
      },
    };
    const wrapper = shallow(<MediaButtonHandler {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
