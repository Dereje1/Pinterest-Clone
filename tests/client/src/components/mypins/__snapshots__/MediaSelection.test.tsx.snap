// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handling different media buttons - AI - generated  1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
    }
  }
>
  <ForwardRef(Button)
    color="info"
    component="label"
    disabled={true}
    onClick={[MockFunction]}
    startIcon={<Memo />}
    sx={
      {
        "margin": "2.3vh",
      }
    }
    variant="contained"
  >
    Generated
  </ForwardRef(Button)>
  <ForwardRef(IconButton)
    disableRipple={true}
    disabled={false}
    onClick={[MockFunction]}
  >
    <Memo(ForwardRef(RestartAltIcon)) />
  </ForwardRef(IconButton)>
  <ForwardRef(Typography)>
    2
    /5
  </ForwardRef(Typography)>
</div>
`;

exports[`Handling different media buttons - AI - not generated yet 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
    }
  }
>
  <ForwardRef(Button)
    color="info"
    component="label"
    disabled={false}
    onClick={[MockFunction]}
    startIcon={<Memo />}
    sx={
      {
        "margin": "2.3vh",
      }
    }
    variant="contained"
  >
    Generate Image
  </ForwardRef(Button)>
  <ForwardRef(IconButton)
    disableRipple={true}
    disabled={true}
    onClick={[MockFunction]}
  >
    <Memo(ForwardRef(RestartAltIcon)) />
  </ForwardRef(IconButton)>
  <ForwardRef(Typography)>
    2
    /5
  </ForwardRef(Typography)>
</div>
`;

exports[`Handling different media buttons - links 1`] = `
<ForwardRef(TextField)
  color="success"
  error={false}
  id="pin-img-link"
  label="Paste image address here http://..."
  onChange={[MockFunction]}
  style={
    {
      "margin": "1.5vh",
      "width": "100%",
    }
  }
  value=""
  variant="standard"
/>
`;

exports[`Handling different media buttons - uploads - with no error 1`] = `
<ForwardRef(Button)
  color="secondary"
  component="label"
  startIcon={<Memo />}
  sx={
    {
      "margin": "2.3vh",
    }
  }
  variant="contained"
>
  replace image
  <input
    accept="image/*"
    hidden={true}
    onChange={[MockFunction]}
    type="file"
  />
</ForwardRef(Button)>
`;

exports[`Handling different media buttons - uploads with error 1`] = `
<ForwardRef(Button)
  color="error"
  component="label"
  startIcon={<Memo />}
  sx={
    {
      "margin": "2.3vh",
    }
  }
  variant="contained"
>
  choose image
  <input
    accept="image/*"
    hidden={true}
    onChange={[MockFunction]}
    type="file"
  />
</ForwardRef(Button)>
`;

exports[`Selecting media will display the media selection toggle buttons 1`] = `
<ForwardRef(ToggleButtonGroup)
  aria-label="text alignment"
  exclusive={true}
  onChange={[MockFunction]}
  sx={
    {
      "color": "900",
      "marginRight": 80,
    }
  }
  value="link"
>
  <ForwardRef(ToggleButton)
    aria-label="link type"
    value="link"
  >
    <Memo(ForwardRef(LinkIcon)) />
  </ForwardRef(ToggleButton)>
  <ForwardRef(ToggleButton)
    aria-label="upload type"
    value="upload"
  >
    <Memo(ForwardRef(DriveFolderUploadIcon)) />
  </ForwardRef(ToggleButton)>
  <ForwardRef(ToggleButton)
    aria-label="AI type"
    disabled={false}
    value="AI"
  >
    <Memo(ForwardRef(PsychologyIcon)) />
  </ForwardRef(ToggleButton)>
</ForwardRef(ToggleButtonGroup)>
`;
